[Custom Fields] Differentiate between raw JSON values and String values that contain valid JSON

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14271
<!-- Id number of the GitHub issue this PR addresses. -->

@hafizrahman @selanthiraiyan not urgent, please review when you have some free time.

## Description
As detailed in the above issue, the current implementation in iOS doesn't allow to differentiate between JSON values and String values that contain valid JSON, and this causes a difference with how Android and Core behaves.
This PR updates the MetaData parsing to expose information on whether the value is a String or a Json using an `enum`, and we store this information also when caching: the logic here is simple, we store the information as it's in the response, `Strings` will be stored with their quotes, and JSON values will be stored as they are without quotes. (the implementation is similar to what we [did](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/model/metadata/WCMetaDataValue.kt) in Android, but simpler here)

While the number of affected files may seem big, the changes are quite simple actually:
1. Whenever the value is read, we read it using the property `stringValue`, this returns the same content as before.
2. When we save MetaData to storage, we use the property `rawValue`, this makes sure that String values are stored with their `quotes`.

**Note:** While the impact of these changes could be small for this specific feature, I think it could have some benefits in the future, for example if we move Subscriptions or Addons to use the `MetaData` class directly, then we'll have to make sure we can use JSON data with it.

## Steps to reproduce
1. Open wp-admin and add a custom field with the following content: `{"key":"value"}`
2. Use the API to add a true JSON field like this for example:
```
POST {site}/wc/v3/orders/{orderId}

{
    "meta_data": [
        {
            "key": "json_object",
            "value": {
                "Key": "Value"
            }
        },
        {
            "key": "json_array",
            "value": [
                {
                    "Key": "Value"
                }
            ]
        }
    ]
}
```
4. Open the given product or order where the changes were made.

## Testing information
- Confirm that the value added in wp-admin is editable in the app.
- Confirm that the value added using the API call is not editable.
- Smoke test the feature as a whole.

## Screenshots

https://github.com/user-attachments/assets/13292455-5a29-47d5-b5fb-c463dc9ada72

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hichamboushaba committed:
Introduce logic to differentiate different types of the metadata value

hichamboushaba committed:
Fix unit tests

hichamboushaba committed:
Improve logic of parsing MetaData

hichamboushaba committed:
Update custom fields to allow editing string wrapped JSON values

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on Networking/Networking/Model/Copiable/Models+Copiable.generated.swift:
```diff
@@ -847,7 +847,7 @@ extension Networking.GoogleAdsCampaign {
         rawType: CopiableProp<String> = .copy,
         amount: CopiableProp<Double> = .copy,
-        country: CopiableProp<String> = .copy,
+        country: NullableCopiableProp<String> = .copy,
```
I think the file wasn't re-generated after some changes in a previous PR, that's why `rage generate` updated it.

hichamboushaba committed:
Updates to ensure MetaData is encoded correctly in all cases

Also ensures we can't bypass the validation by instantiating the enum case directly.

hichamboushaba commented on Networking/Networking/Model/MetaData.swift:
```diff
@@ -39,18 +47,92 @@ private extension MetaData {
-                return String(data: data, encoding: .utf8) ?? ""
-            } else {
-                return "\(value)"
+public struct MetaDataValue: Codable, Equatable, Sendable {
```
We expose a `struct` instead of exposing the enum directly just to make sure we have some validation, and we no consumer can bypass it by instantiating the enum directly, ie: `.json("invalid json")`, this is actually because as far as I found, Swift doesn't allow to have private initializer for the enum cases (unlike Kotlin with its `sealed interface` feature).

hafizrahman replied:
I think this is a good architectural choice in terms of replicating `sealed interface` in Kotlin.

With that being said, I'm curious if Swift has a commonly used alternative for this purpose. cc @selanthiraiyan for opinion.

selanthiraiyan replied:
> We expose a struct instead of exposing the enum directly just to make sure we have some validation, and we no consumer can bypass it by instantiating the enum directly, ie: .json("invalid json")

I hope you are referring to the `ValueHolder` enum here. I don't know of any way to prevent directly initialising an enum with an associated value. 

The `MetaDataValue` wrapper technique looks good to me.

hichamboushaba commented on Networking/Networking/Model/MetaData.swift:
```diff
@@ -39,18 +47,92 @@ private extension MetaData {
+private extension MetaDataValue {
+    private enum ValueHolder: Codable, Equatable {
+        case string(_ value: String)
+        case json(_ json: String)
```
I played with a different approach here, instead of making the backing data a `String`, we can have it as `Codable` directly, I don't have a strong opinion on which approach is best, so I kept the `String` based implementation.

```swift
public struct MetaDataValue: Codable, Equatable, Sendable {
    private let valueHolder: ValueHolder

    public var stringValue: String {
        valueHolder.stringValue
    }
    public var rawValue: String {
        valueHolder.rawValue
    }

    public var isJson: Bool {
        switch valueHolder {
        case .string:
            return false
        case .json:
            return true
        }
    }

    public init(rawValue: String) {
        let data = Data(rawValue.utf8)
        if let jsonObject = try? JSONDecoder().decode(AnyCodable.self, from: data) {
            valueHolder = try! ValueHolder(from: jsonObject.value)
        } else {
            valueHolder = .string(rawValue.removingPrefix("\"").removingSuffix("\""))
        }
    }

    public init(from decoder: Decoder) throws {
        valueHolder = try ValueHolder(from: decoder)
    }

    public func encode(to encoder: Encoder) throws {
        try valueHolder.encode(to: encoder)
    }
}

private extension MetaDataValue {
    private enum ValueHolder: Codable, Equatable {
        case string(_ value: String)
        case json(_ json: Codable)

        var stringValue: String {
            switch self {
            case .string(let value):
                return value
            case .json(let json):
                return json.asJSONString()
            }
        }

        var rawValue: String {
            switch self {
            case .string(let value):
                return "\"\(value)\""
            case .json(let json):
                return json.asJSONString()
            }
        }

        init(from decoder: any Decoder) throws {
            let container = try decoder.singleValueContainer()
            let value = try container.decode(AnyCodable.self)
            try self.init(from: value.value)
        }

        init(from object: Any) throws {
            switch object {
            case let value as String:
                self = .string(value)
            case is Bool:
                self = .string(String(describing: object))
            case is any Numeric:
                self = .string(String(describing: object))
            default:
                self = .json(AnyCodable(object))
            }
        }

        func encode(to encoder: Encoder) throws {
            switch self {
            case .string(let value):
                try value.encode(to: encoder)
            case .json(let json):
                try json.encode(to: encoder)
            }
        }

        static func == (lhs: ValueHolder, rhs: ValueHolder) -> Bool {
            return lhs.stringValue == rhs.stringValue
        }
    }
}

private extension Encodable {
    func asJSONString() -> String {
        if let data = try? JSONEncoder().encode(self) {
            return String(data: data, encoding: .utf8) ?? ""
        }
        return ""
    }
}
```

hafizrahman replied:
Interesting alternative approach with `Codable`. My hunch here is that for the current use case where we only need to validate if a value is valid JSON and display it in a read-only field, the String approach is more straightforward.

hafizrahman commented on Networking/Networking/Model/MetaData.swift:
The comment
```
/// Currently only handles `String` metadata values
```

on line 5 will need updating with this change, I believe?

hichamboushaba replied:
Good catch, I'll make sure to remove it before merging.

hafizrahman commented on Networking/Networking/Model/MetaData.swift:
```diff
@@ -39,18 +47,92 @@ private extension MetaData {
+    public init(rawValue: String) {
+        let data = Data(rawValue.utf8)
+        if let jsonObject = try? JSONDecoder().decode(AnyCodable.self, from: data) {
+            valueHolder = try! ValueHolder(from: jsonObject.value)
```
We generally want to avoid force unwrapping: https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/coding-style-guide.md#forced-downcasts-and-unwrapping

hichamboushaba replied:
I may be missing something, so please bear with me here 😅.

I don't think I'm using any downcasting or unwrapping here, the `try!` is needed just to make the compiler happy as the `ValueHolder.init` is marked as `throws`: This should never crash, as the only case where `ValueHolder.init` could throw is it fails to encode the content as JSON, but since we confirm that the content has a valid JSON here, it should never throw.

hafizrahman replied:
What @selanthiraiyan [said here](https://github.com/woocommerce/woocommerce-ios/pull/14292/files#r1837494558) was what I had in mind.

hichamboushaba replied:
I applied that change, thank you both.

But just for transparency, and as I shared with @selanthiraiyan, I personally prefer to use `try!` here, as having a crash (if for example the `ValueHoder.init` was modifier later) is better than using incorrect data.

hafizrahman commented on Networking/Networking/Model/MetaData.swift:
```diff
@@ -23,9 +31,9 @@ public struct MetaData: Codable, Equatable, Sendable, GeneratedCopiable, Generat
         let metadataID = try container.decode(Int64.self, forKey: .metadataID)
         let key = try container.decode(String.self, forKey: .key)
-        let rawValue = container.failsafeDecodeIfPresent(AnyCodable.self, forKey: .value) ?? ""
+        let value = try MetaDataValue(from: container.superDecoder(forKey: CodingKeys.value))
```
Nice use of `superDecoder`, TIL for me.

hafizrahman reviewed (commented):
Sorry I focused on other tasks last week before reviewing this. In general this looks quite robust to me. I left some comments.

selanthiraiyan commented on Networking/Networking/Model/MetaData.swift:
```diff
@@ -39,18 +47,92 @@ private extension MetaData {
+
+    public init(rawValue: String) {
+        let data = Data(rawValue.utf8)
+        if let jsonObject = try? JSONDecoder().decode(AnyCodable.self, from: data) {
+            valueHolder = try! ValueHolder(from: jsonObject.value)
+        } else {
+            valueHolder = .string(rawValue.removingPrefix("\"").removingSuffix("\""))
+        }
```
Sorry to jump into the discussion. I would suggest using the following to avoid force unwrapping. 

```suggestion
        if let jsonObject = try? JSONDecoder().decode(AnyCodable.self, from: data),
           let value = try? ValueHolder(from: jsonObject.value) {
            valueHolder = value
        } else {
            valueHolder = .string(rawValue.removingPrefix("\"").removingSuffix("\""))
        }
```

hichamboushaba committed:
Update code to avoid force unwrapping

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

hichamboushaba committed:
Remove outdated comment

hichamboushaba committed:
Merge branch 'trunk' into issue/14271-better-json-detection

hafizrahman reviewed (approved):
Thanks for the update, this LGTM.

hichamboushaba committed:
Merge branch 'trunk' into issue/14271-better-json-detection

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/232">21.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
